### PR TITLE
Parallelize multiple queries and data updates

### PR DESCRIPTION
**Pull request by NinoFloris**
This PR introduces real query concurrency in the scope of a single http request, not unlike the query pipelining seen in drogon and lithium. Instead of being confined to a single connection like pipelining (which does have its upsides when you have end to end request/core affinity) we spread this load across all connections in the connection pool through multiplexing.

Data updates now calls into this new implementation before pushing its subsequent update.

Estimates based on 'very reproducible benchmarks™️' _compliments of arm64 macOS montery_ tell me we could expect at least 80% more perf on multiple queries and data updates. Should give a nice boost to the TE composite score! 

/cc @roji

Might nicely contribute to https://github.com/dotnet/efcore/issues/26796

EDIT: bringing the connection count for my local wrk up to 256 or 512 brings main and this PR closer for me, but I still see a 60% increase locally.

### CODE REVIEW

**Comment by sebastienros** [comment]
This is real multipelxing. But how I read the code is that it create `count` connections, and there can be n (equals to httpconnections) concurrent requests. So should the pool be configured differently? Doesn't it create too many connections in this case? Should it be capped?

**Comment by roji** [comment]
@sebastienros we may want to see the impact of tweaking the total number of physical connections (Max Pool Size), we already went through a phase like this with @NinoFloris at some point, we'll probably need to do it again. But multiplexing generally manages to push a very large load through few connections. We'll see what the counters tell us.

In any case, I'm off to do a perf run or two on this!

**Comment by NinoFloris** [comment]
@sebastienros these connections are really quite fake and just serve to add another work item to the multiplexing queue (as you can't have concurrent work on an ado.net connection), we're still capped by the pool when it comes to the maximum amount of physical connections we open!

**Comment by sebastienros** [comment]
> just serve to add another work item to the multiplexing queue

I didn't realize it was working this way, clever! So that's specific to npgsql since this is how it will handle "pipelining"? Another driver might choke on this pattern?

**Comment by roji** [comment]
> I didn't realize it was working this way, clever! So that's specific to npgsql since this is how it will handle "pipelining"? Another driver might choke on this pattern?

Absolutely :) When multiplexing is on, an NpgsqlConnection no longer corresponds to a physical connection. It's just there because ADO.NET wants it, but calling NpgsqlConnection.Open is basically a no-op.

**Comment by roji** [comment]
BTW @sebastienros I have some ideas for a new ADO.NET API which would make this all much nicer.

**Comment by NinoFloris** [comment]
Absolutely, the underlying support for multiplexing is key here. 

I initially started with a version that spread the load over `ProcessorCount` *connection pools* with a max pool size of 1 to really simulate pipelining. It didn't seem quite worth it due to the distribution key for which connection pool to pick being hard to pick fairly (kestrel connection id may work though). Just doing `managedthreadid % processorcount` lead to a less than ideal db work distribution.

Leaving that issue aside we also don't have any affinity for task continuations to keep work from shifting around. I expect solving that before doing pipelining makes more sense. After that pipelining would be a natural improvement in the cases where - at high core counts and tiny work items - the single multiplexing write thread in npgsql can't keep up anymore (as I suspect scaling up that thread would bring a lot of contention into the mix, just like the threadpool has to deal with this at higher core counts).

**Comment by roji** [comment]
Since perflab machines are a bit busy, I ran on aspnet-perf-win in the meantime, below are the results. Numbers aren't as good as expected, but they're still very nice and this is a considerably weaker system (plus Windows). I'll do aspnet-citrine-lin too when it frees up.

Before commit: 4f618bf867f818edb888a6ff2c334586735cabfa
After commit: f092baa6a4ea8b16bcb5c5852093923dd94d0d4e

benchmark        | Before RPS | After RPS | Change
---------------- | ---------- | --------- | ------
Multiple queries | 10,849     | 12,877    | +18.6% 
Multiple updates | 8,331      | 9,180     | +10.1%

<details>
<summary>Detailed run results</summary>

### Multiple queries before

```
crank --config https://raw.githubusercontent.com/aspnet/Benchmarks/main/scenarios/platform.benchmarks.yml --scenario multiple_queries --profile aspnet-perf-win --application.framework net6.0

| db                  |         |
| ------------------- | ------- |
| CPU Usage (%)       | 92      |
| Cores usage (%)     | 1,102   |
| Working Set (MB)    | 44      |
| Build Time (ms)     | 2,092   |
| Start Time (ms)     | 424     |
| Published Size (KB) | 914,425 |


| application           |               |
| --------------------- | ------------- |
| CPU Usage (%)         | 52            |
| Cores usage (%)       | 626           |
| Working Set (MB)      | 292           |
| Private Memory (MB)   | 429           |
| Build Time (ms)       | 3,633         |
| Start Time (ms)       | 3,065         |
| Published Size (KB)   | 96,714        |
| .NET Core SDK Version | 6.0.100       |
| ASP.NET Core Version  | 6.0.0+ae1a6cb |
| .NET Runtime Version  | 6.0.0+4822e3c |


| load                   |         |
| ---------------------- | ------- |
| CPU Usage (%)          | 5       |
| Cores usage (%)        | 58      |
| Working Set (MB)       | 38      |
| Private Memory (MB)    | 358     |
| Start Time (ms)        | 0       |
| First Request (ms)     | 100     |
| Requests/sec           | 10,849  |
| Requests               | 163,622 |
| Mean latency (ms)      | 23.68   |
| Max latency (ms)       | 97.74   |
| Bad responses          | 0       |
| Socket errors          | 0       |
| Read throughput (MB/s) | 7.84    |
| Latency 50th (ms)      | 21.30   |
| Latency 75th (ms)      | 26.87   |
| Latency 90th (ms)      | 34.09   |
| Latency 99th (ms)      | 48.30   |
```

### Multiple queries after

```
crank --config https://raw.githubusercontent.com/aspnet/Benchmarks/main/scenarios/platform.benchmarks.yml --scenario multiple_queries --profile aspnet-perf-win --application.framework net6.0 --application.source.repository http://github.com/NinoFloris/Benchmarks --application.source.branchOrCommit enhancement/aspnetcore-multiple-queries

| db                  |         |
| ------------------- | ------- |
| CPU Usage (%)       | 99      |
| Cores usage (%)     | 1,192   |
| Working Set (MB)    | 44      |
| Build Time (ms)     | 1,953   |
| Start Time (ms)     | 448     |
| Published Size (KB) | 914,425 |


| application           |               |
| --------------------- | ------------- |
| CPU Usage (%)         | 33            |
| Cores usage (%)       | 394           |
| Working Set (MB)      | 476           |
| Private Memory (MB)   | 493           |
| Build Time (ms)       | 3,631         |
| Start Time (ms)       | 2,400         |
| Published Size (KB)   | 96,714        |
| .NET Core SDK Version | 6.0.100       |
| ASP.NET Core Version  | 6.0.0+ae1a6cb |
| .NET Runtime Version  | 6.0.0+4822e3c |


| load                   |         |
| ---------------------- | ------- |
| CPU Usage (%)          | 6       |
| Cores usage (%)        | 77      |
| Working Set (MB)       | 38      |
| Private Memory (MB)    | 358     |
| Start Time (ms)        | 0       |
| First Request (ms)     | 90      |
| Requests/sec           | 12,877  |
| Requests               | 194,202 |
| Mean latency (ms)      | 19.88   |
| Max latency (ms)       | 46.04   |
| Bad responses          | 0       |
| Socket errors          | 0       |
| Read throughput (MB/s) | 9.30    |
| Latency 50th (ms)      | 19.77   |
| Latency 75th (ms)      | 22.26   |
| Latency 90th (ms)      | 24.49   |
| Latency 99th (ms)      | 30.27   |
```

### Multiple updates before

```
crank --config https://raw.githubusercontent.com/aspnet/Benchmarks/main/scenarios/platform.benchmarks.yml --scenario updates --profile aspnet-perf-win --application.framework net6.0

| db                  |         |
| ------------------- | ------- |
| CPU Usage (%)       | 94      |
| Cores usage (%)     | 1,132   |
| Working Set (MB)    | 50      |
| Build Time (ms)     | 2,742   |
| Start Time (ms)     | 409     |
| Published Size (KB) | 914,425 |


| application           |               |
| --------------------- | ------------- |
| CPU Usage (%)         | 48            |
| Cores usage (%)       | 570           |
| Working Set (MB)      | 321           |
| Private Memory (MB)   | 431           |
| Build Time (ms)       | 2,457         |
| Start Time (ms)       | 2,334         |
| Published Size (KB)   | 96,714        |
| .NET Core SDK Version | 6.0.100       |
| ASP.NET Core Version  | 6.0.0+ae1a6cb |
| .NET Runtime Version  | 6.0.0+4822e3c |


| load                   |         |
| ---------------------- | ------- |
| CPU Usage (%)          | 4       |
| Cores usage (%)        | 47      |
| Working Set (MB)       | 38      |
| Private Memory (MB)    | 358     |
| Start Time (ms)        | 0       |
| First Request (ms)     | 110     |
| Requests/sec           | 8,331   |
| Requests               | 125,798 |
| Mean latency (ms)      | 30.75   |
| Max latency (ms)       | 99.16   |
| Bad responses          | 0       |
| Socket errors          | 0       |
| Read throughput (MB/s) | 6.02    |
| Latency 50th (ms)      | 28.95   |
| Latency 75th (ms)      | 35.02   |
| Latency 90th (ms)      | 42.58   |
| Latency 99th (ms)      | 57.45   |
```

### Multiple updates after

```
crank --config https://raw.githubusercontent.com/aspnet/Benchmarks/main/scenarios/platform.benchmarks.yml --scenario updates --profile aspnet-perf-win --application.framework net6.0 --application.source.repository http://github.com/NinoFloris/Benchmarks --application.source.branchOrCommit enhancement/aspnetcore-multiple-queries

| db                  |         |
| ------------------- | ------- |
| CPU Usage (%)       | 94      |
| Cores usage (%)     | 1,127   |
| Working Set (MB)    | 50      |
| Build Time (ms)     | 1,942   |
| Start Time (ms)     | 413     |
| Published Size (KB) | 914,425 |


| application           |               |
| --------------------- | ------------- |
| CPU Usage (%)         | 30            |
| Cores usage (%)       | 362           |
| Working Set (MB)      | 407           |
| Private Memory (MB)   | 540           |
| Build Time (ms)       | 2,310         |
| Start Time (ms)       | 2,798         |
| Published Size (KB)   | 96,714        |
| .NET Core SDK Version | 6.0.100       |
| ASP.NET Core Version  | 6.0.0+ae1a6cb |
| .NET Runtime Version  | 6.0.0+4822e3c |


| load                   |         |
| ---------------------- | ------- |
| CPU Usage (%)          | 5       |
| Cores usage (%)        | 61      |
| Working Set (MB)       | 38      |
| Private Memory (MB)    | 358     |
| Start Time (ms)        | 0       |
| First Request (ms)     | 106     |
| Requests/sec           | 9,180   |
| Requests               | 138,514 |
| Mean latency (ms)      | 27.91   |
| Max latency (ms)       | 49.14   |
| Bad responses          | 0       |
| Socket errors          | 0       |
| Read throughput (MB/s) | 6.63    |
| Latency 50th (ms)      | 27.91   |
| Latency 75th (ms)      | 30.45   |
| Latency 90th (ms)      | 32.86   |
| Latency 99th (ms)      | 38.22   |
```

</details>

**Comment by sebastienros** [comment]
If only we had a tool to generate comparisons between runs.

File: scenarios/platform.benchmarks.yml [hunk 1]
```diff
@@ -17,8 +17,10 @@ jobs:
     readyStateText: Application started.
     framework: net5.0
     environmentVariables:
+      DOTNET_SYSTEM_NET_SOCKETS_INLINE_COMPLETIONS: 1
```
Comment (sebastienros) [inline scenarios/platform.benchmarks.yml:20]: It's fine for testing in a PR, but we don't want to run this benchmark by default with this setting. the idea is that we'll add more scenarios that represent the absolute best tuning just for numbers. We want to track the perf of the system without such tuning to detect changes that impact perf. But some discussions I had with Shay showed that we still need to have some benchmarks that represent exactly what we have or will have on TE.
Another example is that this one doesn't have PGO, but we do have some scenarios with only PGO and no inlining, or just inlining and no PGO. TE's repository has both.

File: scenarios/platform.benchmarks.yml [hunk 2]
```diff
@@ -17,8 +17,10 @@ jobs:
...
+      DOTNET_SYSTEM_NET_SOCKETS_THREAD_COUNT: 1
       database: PostgreSQL
-      connectionstring: Server={{databaseServer}};Database=hello_world;User Id=benchmarkdbuser;Password=benchmarkdbpass;Maximum Pool Size=28;Enlist=false;Max Auto Prepare=4;Multiplexing=true;Write Coalescing Buffer Threshold Bytes=1000
+      connectionstring: Server={{databaseServer}};Database=hello_world;User Id=benchmarkdbuser;Password=benchmarkdbpass;Maximum Pool Size=20;Enlist=false;Max Auto Prepare=4;Multiplexing=true;Write Coalescing Buffer Threshold Bytes=1400
```
Comment (sebastienros) [inline scenarios/platform.benchmarks.yml:23]: Would love to see some comments with how these numbers were chosen so we have an idea on how to check if they are still the most appropriate.

**Comment by NinoFloris** [comment]
@sebastienros I wouldn't bother reviewing this too closely, for the time being it's a convenient mechanism to ship @roji changes to benchmark in the perflab.

This last commit is one big hack to test some assumptions, not fit for merging at all. 

Locally on my M1 Max I'm getting close to 18k rps while running the db and wrk on the same machine. I'd really like to figure out why we don't see similar gains in the perflab...

**Comment by sebastienros** [comment]
@NinoFloris perfect then, good luck!

**Comment by NinoFloris** [comment]
Closing until we know which way to go, experiments gave us a lot of good directions though!

**Pull request closed without merging**